File: frontend/src/types/crm.ts
export interface Account {
  id: string
  name: string
  industry?: string
  phone?: string
  website?: string
  email?: string
  billing_street?: string
  billing_city?: string
  billing_state?: string
  billing_zip?: string
  billing_country?: string
  description?: string
  owner_id?: string
  created_at: string
  updated_at: string
}

export interface AccountCreate {
  name: string
  industry?: string
  phone?: string
  website?: string
  email?: string
  billing_street?: string
  billing_city?: string
  billing_state?: string
  billing_zip?: string
  billing_country?: string
  description?: string
  owner_id?: string
}

export interface AccountUpdate extends Partial<AccountCreate> {}

export interface Contact {
  id: string
  first_name: string
  last_name: string
  email?: string
  phone?: string
  mobile_phone?: string
  title?: string
  department?: string
  account_id?: string
  account_name?: string
  accounts?: ContactAccount[]
  owner_id?: string
  created_at: string
  updated_at: string
}

export interface ContactCreate {
  first_name: string
  last_name: string
  email?: string
  phone?: string
  mobile_phone?: string
  title?: string
  department?: string
  account_id?: string
  owner_id?: string
}

export interface ContactUpdate extends Partial<ContactCreate> {}

export interface ContactAccount {
  id: string
  contact_id: string
  account_id: string
  account_name?: string
  assigned_at?: string
}

export interface ContactAccountCreate {
  account_id: string
}

export interface Product {
  id: string
  name: string
  product_code?: string
  description?: string
  price?: number
  cost?: number
  category?: string
  is_active?: boolean
  owner_id?: string
  created_at: string
  updated_at: string
}

export interface ProductCreate {
  name: string
  product_code?: string
  description?: string
  price?: number
  cost?: number
  category?: string
  is_active?: boolean
  owner_id?: string
}

export interface ProductUpdate extends Partial<ProductCreate> {}

export interface Stage {
  id: string
  name: string
  probability: number
  sort_order: number
  is_closed_won: boolean
  is_closed_lost: boolean
}

export interface Opportunity {
  id: string
  name: string
  account_id?: string
  stage_id: string
  amount?: number
  probability?: number
  close_date?: string
  description?: string
  owner_id?: string
  created_at: string
  updated_at: string
}

export interface OpportunityCreate {
  name: string
  account_id?: string
  stage_id: string
  amount?: number
  probability?: number
  close_date?: string
  description?: string
  owner_id?: string
}

export interface OpportunityUpdate extends Partial<OpportunityCreate> {}

export interface PipelineStageData {
  stage: Stage
  opportunities: Opportunity[]
  total_amount: number
  count: number
}

export interface PipelineData {
  stages: PipelineStageData[]
}

export interface PaginatedResponse<T> {
  total: number
  page: number
  page_size: number
  items: T[]
}

export interface LineItem {
  id: string
  opportunity_id: string
  product_id: string
  quantity: number
  unit_price: number
  total_price: number
  created_at: string
}

export interface LineItemCreate {
  product_id: string
  quantity: number
  unit_price: number
}

// ── Custom Objects ────────────────────────────────────────────────

export interface CustomFieldDef {
  id: string
  object_id: string
  api_name: string
  label: string
  field_type: string
  is_required: boolean
  is_unique: boolean
  default_value?: string | null
  max_length?: number | null
  picklist_values?: string[] | null
  precision_total?: number | null
  precision_scale?: number | null
  lookup_object_id?: string | null
  display_order: number
  created_at: string
  updated_at: string
}

export interface CustomObjectDef {
  id: string
  api_name: string
  label: string
  plural_label?: string
  description?: string
  table_name: string
  is_active: boolean
  fields: CustomFieldDef[]
  created_at: string
  updated_at: string
}

export interface CustomFieldCreate {
  api_name: string
  label: string
  field_type: string
  is_required?: boolean
  is_unique?: boolean
  default_value?: string | null
  max_length?: number | null
  picklist_values?: string[] | null
  precision_total?: number | null
  precision_scale?: number | null
  lookup_object_id?: string | null
  display_order?: number
}

export interface CustomObjectCreate {
  api_name: string
  label: string
  plural_label?: string
  description?: string
  fields: CustomFieldCreate[]
}

export interface CustomRecord {
  id: string
  record_id: string
  owner_id?: string | null
  fields: Record<string, any>
  created_at: string
  updated_at: string
}

File: backend/app/schemas/crm.py
from pydantic import BaseModel, model_validator
from datetime import datetime, date
from typing import Optional, Any


class AccountBase(BaseModel):
    name: str
    industry: Optional[str] = None
    phone: Optional[str] = None
    website: Optional[str] = None
    email: Optional[str] = None
    billing_street: Optional[str] = None
    billing_city: Optional[str] = None
    billing_state: Optional[str] = None
    billing_zip: Optional[str] = None
    billing_country: Optional[str] = None
    description: Optional[str] = None
    owner_id: Optional[str] = None


class AccountCreate(AccountBase):
    pass


class AccountUpdate(AccountBase):
    name: Optional[str] = None


class AccountOut(AccountBase):
    id: str
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class ContactBase(BaseModel):
    first_name: str
    last_name: str
    email: Optional[str] = None
    phone: Optional[str] = None
    mobile_phone: Optional[str] = None
    title: Optional[str] = None
    department: Optional[str] = None
    account_id: Optional[str] = None
    owner_id: Optional[str] = None


class ContactCreate(ContactBase):
    pass


class ContactUpdate(ContactBase):
    first_name: Optional[str] = None
    last_name: Optional[str] = None


class ContactOut(ContactBase):
    id: str
    account_name: Optional[str] = None
    accounts: list["ContactAccountOut"] = []
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True

    @model_validator(mode="before")
    @classmethod
    def _map_account_associations(cls, data):
        """Map ORM model's account_associations to schema's accounts field."""
        if isinstance(data, dict) and "account_associations" in data:
            if "accounts" not in data or not data["accounts"]:
                data["accounts"] = data.pop("account_associations")
        return data


class ContactAccountOut(BaseModel):
    id: str
    contact_id: str
    account_id: str
    account_name: Optional[str] = None
    assigned_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class ContactAccountCreate(BaseModel):
    account_id: str


class StageOut(BaseModel):
    id: str
    name: str
    probability: int
    sort_order: int
    is_closed_won: bool
    is_closed_lost: bool

    class Config:
        from_attributes = True


class OpportunityBase(BaseModel):
    name: str
    account_id: Optional[str] = None
    stage_id: str
    amount: Optional[float] = 0.0
    probability: Optional[int] = 0
    close_date: Optional[date] = None
    description: Optional[str] = None
    owner_id: Optional[str] = None


class OpportunityCreate(OpportunityBase):
    pass


class OpportunityUpdate(BaseModel):
    name: Optional[str] = None
    account_id: Optional[str] = None
    stage_id: Optional[str] = None
    amount: Optional[float] = None
    probability: Optional[int] = None
    close_date: Optional[date] = None
    description: Optional[str] = None
    owner_id: Optional[str] = None


class OpportunityProductBase(BaseModel):
    product_id: str
    quantity: int = 1
    unit_price: float = 0.0
    total_price: float = 0.0


class OpportunityProductCreate(OpportunityProductBase):
    pass


class OpportunityProductOut(OpportunityProductBase):
    id: str
    opportunity_id: str
    created_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class OpportunityOut(OpportunityBase):
    id: str
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    line_items: list[OpportunityProductOut] = []

    class Config:
        from_attributes = True


class PipelineStageData(BaseModel):
    stage: StageOut
    opportunities: list[OpportunityOut]
    total_amount: float
    count: int


class PipelineOut(BaseModel):
    stages: list[PipelineStageData]


class ProductBase(BaseModel):
    name: str
    product_code: Optional[str] = None
    description: Optional[str] = None
    price: Optional[float] = 0.0
    cost: Optional[float] = 0.0
    category: Optional[str] = None
    is_active: Optional[bool] = True
    owner_id: Optional[str] = None


class ProductCreate(ProductBase):
    pass


class ProductUpdate(ProductBase):
    name: Optional[str] = None


class ProductOut(ProductBase):
    id: str
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class PaginatedResponse(BaseModel):
    total: int
    page: int
    page_size: int
    items: list[Any]
